labman/db/plate.py

```diff
@@ -300,3 +390,59 @@ def get_well(self, row, column):
                     % (row, column, self.id))
 
             return container_module.Well(res[0][0])
+
+    def get_wells_by_sample(self, sample_id):
+        """Returns the list of wells containing the given sample
+
+        Parameters
+        ----------
+        sample_id : str
+            The sample to search for
+
+        Returns
+        -------
+        list of labman.db.container.Well
+        """
+        with sql_connection.TRN as TRN:
+            sql = """SELECT well_id
+                     FROM qiita.well
+                        JOIN qiita.composition USING (container_id)
+                        JOIN qiita.sample_composition USING (composition_id)
+                     WHERE plate_id = %s AND sample_id = %s
+                     ORDER BY well_id"""
+            TRN.add(sql, [self.id, sample_id])
+            res = [container_module.Well(well)
+                   for well in TRN.execute_fetchflatten()]
+        return res
+
+    def get_previously_plated_wells(self):
+        """Get wells with samples that have been previously plated
+
+        Returns
+        -------
+        dict of {well: list of plates}
+            The wells that contain samples that have been previously plated
+            and the plates in which they're found
+        """
+        with sql_connection.TRN as TRN:
+            sql = """SELECT plate_id, array_agg(sample_id)
+                     FROM qiita.sample_composition
+                        JOIN qiita.composition USING (composition_id)
+                        JOIN qiita.well USING (container_id)
+                     WHERE plate_id <> %s AND sample_id IN (
+                        SELECT sample_id
+                        FROM qiita.sample_composition
+                            JOIN qiita.composition USING (composition_id)
+                            JOIN qiita.well USING (container_id)
+                            WHERE plate_id = %s)
+                     GROUP BY plate_id"""
+            TRN.add(sql, [self.id, self.id])
+            prev_plated = TRN.execute_fetchindex()
```

Comment:
Same here. A little more complex SQL will avoid multiple queries.

Reply:
In this case that is true, it is performing multiple queries.

However, I've not been able to come up with a SQL query that achieves what I want here - I'm open for ideas.

Reply:
K, let's leave it like this and we can improve later if it becomes a bottleneck.